packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -1327,23 +1327,14 @@ function applyAttributes(ctx: ParserCtx, parsedAttr: ParsedAttribute, element: B
             );
         }
 
-        if (!/^-*[a-z]/.test(name)) {
-            ctx.throwOnNode(
-                ParserDiagnostics.ATTRIBUTE_NAME_MUST_START_WITH_ALPHABETIC_OR_HYPHEN_CHARACTER,
-                attr,
-                [name, tag]
-            );
-        }
-
-        // disallow attr name which combines underscore character with special character.
-        // We normalize camel-cased names with underscores caMel -> ca-mel; thus sanitization.
-        // Note 1: underscore followed by a hyphen is valid to accomodate property names like thirsty_Camel -> thirsty_-camel
-        if (name.match(/_[^a-z0-9-]|[^a-z0-9]_/)) {
-            ctx.throwOnNode(
-                ParserDiagnostics.ATTRIBUTE_NAME_CANNOT_COMBINE_UNDERSCORE_WITH_SPECIAL_CHARS,
-                attr,
-                [name, tag]
-            );
+        // The leading '-' is necessary to preserve attribute to property reflection as the '-' is a signal
+        // to the compiler to convert the first character following it to an uppercase.
+        // This is needed for property names with an @api annotation because they can begin with an upper case character.
+        if (!/^-*[a-z_$]/.test(name)) {
```

Comment:
I found the reference to why we need the `-` character before the matching criteria in #1687.

I chose to leave the leading `-` because it is a signal to the compiler to capitalize the next characters.

Reply:
Why is it necessary to add the `$` to the regex? I assume we need to support this as well?

Reply:
Yup, the flexipage team needs attribute names to start with `$` as well.  Adding it here to capture any characters that start with `$`.

Although now that I think about it this validation doesn't seem quite right either.  The original validation really validates against 2 scenarios:

1. If the leading character in the attribute name is an alphabetic character.
2. If the leading character in the attribute name is a `-` it must be followed by an alphabetic character.

The error message, in this case, doesn't seem accurate as it only indicates the leading character needs to start with a `-` (not followed by an alphabetic character).  If a Latin character follows the `-` for example, `<x-foo -ñana="haha">` will throw this error.

Is the second restriction necessary?  As far as I can tell in the [capitalization logic](https://github.com/salesforce/lwc/blob/master/packages/@lwc/template-compiler/src/shared/utils.ts#L9), the `toUpperCase` method is called on any character after a `-` is seen.

Depending on what we decide to do this PR should be updated as follows:

1. Keep scenario 2: 
    - Update the error message to include scenario 2 if we keep it and update the regex to `!/^-*[a-z]|^[_$]/.test(name)`.
2. Remove scenarios 2:
    - The regex should be modified to `name.match(/^[^a-z-_$]/)` and only validate against the leading character in the attribute name.

Reply:
> Why is it necessary to add the $ to the regex

It was decided to include it, to be proactive about common symbols that devs are likely to use. The `$` is a valid starting character for a javascript class field. 
The $ character was not actually in the [usage data](https://docs.google.com/spreadsheets/d/14yJbgk7VxIiD56PiGvegjGF4a6JcVqTOM5fuNlTbCCQ/edit#gid=1147187344) collected by flexipage team. They only requested for `_`.

Reply:
> @pmdartus do you have any thoughts on this?

If I understand correctly the question, I would suggest making the least amount of changes to the regex to prevent potential regressions and update the error message to something like:

```
Attribute name must start with an optional hyphen character followed by an alphabetic character, an underscore, or a dollar sign.
```